### PR TITLE
Add grad_scale op

### PR DESCRIPTION
**Pull request by YangZhang4065**
I added a new grad_scale op to scale or inverse gradient in back-propagation.

This op could be helpful if you want to
1, Implement adversarial optimization which requires gradient inverse:
https://arxiv.org/abs/1409.7495
https://arxiv.org/abs/1511.06434

2, Have relatively different learning rates across the graph.

### CODE REVIEW

**Comment by lamblin** [comment]
I'm not opposed to it, but maybe we should have a more general mechanism for defining or redefining the gradient of an Op, rather than a collection of Ops for all the things that we may want to do to gradients.
Scaling seems general enough (as well as blocking and clipping, which we already have), so I'm OK with this PR.

**Comment by f0k** [comment]
> maybe we should have a more general mechanism for defining or redefining the gradient of an Op

+1

This comes up on the Lasagne mailing list every now and then. For guided backprop, I used OpFromGraph and replaced the `grad` method, but that's cumbersome (the subgraph inside OpFromGraph cannot interact with the outside graph any more in optimization, so you need to manually define it for the correct device, for example).
Maybe we can have a simpler way to define an Op from two functions (forward and backward pass). It would create a unique placeholder Op with a `grad` method and register an optimization that replaces the placeholder with the forward pass.

**Comment by lamblin** [comment]
@f0k Agreed, that is indeed what we suggest in #5225.

**Comment by f0k** [comment]
> Agreed, that is indeed what we suggest

Ah, great minds think alike :)

**Comment by nouiz** [comment]
@YangZhang4065 Do you have time to finish this PR? It only need the flake8 fixes I comented aboce. Otherwise, it should be good.

thanks

**Comment by nouiz** [comment]
continued in gh-5399
